gtda/diagrams/features.py

```diff
@@ -485,26 +501,36 @@ def fit(self, X, y=None):
 
         if self.n_coefficients is None:
             self.n_coefficients_ = \
-                np.array([X[i].shape[0] for i in range(len(X))]).max()
-        else:
+                [_subdiagrams(X, [dim]).shape[1]
+                 for dim in self.homology_dimensions_]
+        elif type(self.n_coefficients) == list:
+            if len(self.n_coefficients) != len(self.homology_dimensions_):
+                raise ValueError(f'n_coefficients has been passed as a list '
+                                 f'of length {len(self.n_coefficients)} while '
+                                 f'diagrams in `X` have '
+                                 f'{len(self.homology_dimensions_)} homology '
+                                 f'dimensions.')
             self.n_coefficients_ = self.n_coefficients
+        else:
+            self.n_coefficients_ = \
+                [self.n_coefficients] * len(self.homology_dimensions_)
 
         self._polynomial_function = \
             _implemented_polynomial_recipes[self.polynomial_type]
 
         return self
 
-    def _complex_polynomial(self, X):
-        Xt = np.zeros(2 * self.n_coefficients_, )
+    def _complex_polynomial(self, X, n_coefficients):
+        Xt = np.zeros(2 * n_coefficients, )
         X = X[X[:, 0] != X[:, 1]]
 
         roots = self._polynomial_function(X)
         coefficients = np.poly(roots)
 
-        coefficients = np.array(coefficients[::-1])[1:]
-        dimension = min(self.n_coefficients_, coefficients.shape[0])
+        coefficients = np.array(coefficients[::-1])
+        dimension = min(n_coefficients, coefficients.shape[0])
         Xt[:dimension] = coefficients[:dimension].real
-        Xt[self.n_coefficients_:self.n_coefficients_ + dimension] = \
+        Xt[n_coefficients:n_coefficients + dimension] = \
             coefficients[:dimension].imag
```

Comment:
My understanding is still a bit different (sorry, I made a mistake in my previous comment anyway). Please find the code below and an explanation:

`coefficients = np.array(coefficients)[1:]`

I tried to compile an example and pinpoint two properties that I think should be preserved:
1. If `coefficients.shape[0]>n_coefficients`, keep coefficients corresponding to the largest powers,
2. If `coefficients.shape[0]<n_coefficients`, then "append zeros to coefficients".

Example:
```
def coefficients_from_roots(roots, n_coefficients):
    Xt = np.zeros(2 * n_coefficients, )
    coefficients = np.poly(roots)

    coefficients = np.array(coefficients)[1:]
    dimension = min(n_coefficients, coefficients.shape[0])
    Xt[:dimension] = coefficients[:dimension].real
    Xt[n_coefficients:n_coefficients + dimension] = \
        coefficients[:dimension].imag
    return Xt
```
verifies pt.2, as expected:
```
>>> coefficients_from_roots([0, 1, 2, 3, 4], 3)
array([-10.,  35., -50.,   0.,   0.,   0.])
>>> coefficients_from_roots([1, 2, 3, 4], 3)
array([-10.,  35., -50.,   0.,   0.,   0.])
```

Reply:
Ok I will experiment again a bit on my side and come back to you.

Reply:
Agreed! And changed. Thank you for checking!